examples/qualcomm/llama2/qaihub_runner/io_memory.h

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright (c) Qualcomm Innovation Center, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#pragma once
+
+#include <cstdint>
+#include <future>
+#include <memory>
+#include <queue>
+#include <thread>
+#include <vector>
+
+#include <executorch/extension/module/module.h>
+#include <executorch/runtime/executor/method_meta.h>
+
+namespace torch {
+namespace executor {
+
+class Memory {
+ public:
+  Memory(
+      const std::vector<std::string>& pos_embs_path,
+      std::vector<std::shared_ptr<Module>>& modules);
+  virtual ~Memory();
+  virtual void prepare_io(
+      const std::vector<Result<MethodMeta>>& methods_meta) = 0;
+  virtual void update_io(
+      int64_t cur_token,
+      int64_t pos,
+      std::vector<std::vector<Tensor>>& output_tensors) = 0;
+  void* get_mutable_ptr();
+  std::vector<Tensor> get_input_tensors(int shard_index);
+  std::vector<Tensor> get_output_tensors(int shard_index);
+
+ protected:
+  std::unique_ptr<void, std::function<void(void*)>> data_ptr_;
```

Comment:
this doesn't need to use a std::function since both of the lambdas you pass are ordinary functions; can just be `void(*)(void*)`

Reply:
I don't think we can, `unique_ptr` requires deleter to be FunctionObject.